### DISCUSSION TITLE
Can TiTiler replace an XYZ tile server?

### DISCUSSION THREAD

**Discussion by reyemtm** [developmentseed/titiler#377] [Q&A] [answered]
I have tested TiTiler on a small machine and am getting very slow performance (1-2 seconds per ortho map tile), but the machine is underpowered...however TiTiler does not seem to peg the CPU. For ortho raster tiles served from mbtiles, speeds to the client can be under 100ms with a high speed connection, and usually under 50ms. The drawback here is that the mbtiles file is about 3-4x times the size of the COG if not bigger for the same area. The 50ms benchmark is about where Google's imagery comes in at so it seems like a good baseline.

Is TiTiler able to reach these speeds for say a 100GB COG with three bands, or is that just not possible? Regardless, has anyone published benchmarks for various virtual hardware setups and tile serving scenarios? I would like to know if performance any where close to this is possible before I go spinning up a Kubernetes cluster. My guess right now is that the amount of CPU needed to reach these speeds will cost more then the cost for added storage space to store mbtiles on block storage.

**Comment by kylebarron** [comment]
> tested TiTiler on a small machine and am getting very slow performance (1-2 seconds per ortho map tile)

> For ortho raster tiles served from mbtiles, speeds to the client can be under 100ms with a high speed connection, and usually under 50ms

You're testing two completely different things. An mbtiles file is:

- Pre-projected to Web Mercator
- Pre-sliced into XYZ tiles
- A static file

That will always be much faster than dynamic tiling, where that entire process happens on the fly instead of being pre-computed. But imagine you have a data source that changes, say, every hour. It would be crazy (assuming some large area) to create a new mbtiles file for every data change every hour. The point of dynamic tiling is to have flexibility in what images are chosen and displayed. If you don't care about that flexibility, a static dataset like mbtiles will always be faster.

**Reply by kylebarron** [reply]
Re-reading your question, it seems that you're using a single COG at a time, and not mosaicking from multiple files? If so, there are some slight nuances to my answer above. Specifically, if you pre-project your COG into Web Mercator and align the internal blocks to the XYZ grid (what [rio-cogeo calls `web-optimized`](https://cogeotiff.github.io/rio-cogeo/CLI/#create)) then you can get performance very very close to a mbtiles file, because reprojection and slicing is no longer needed. At that point you need a max of 2 HTTP requests to fetch the tile data

- 1 for the header information from the COG (If not cached)
- 1 for the tile data (Akin to the `SELECT` query within the mbtiles file)

At that point there should only be differences depending on where you store the COG vs MBtiles file (e.g. COGs could be on block storage but MBtiles need to be on a file system, which could be faster).

**Reply by reyemtm** [reply]
Is it possible to create a web-optimized COG from `gdal_translate -of COG`, or is there a way to test if a COG is web optimized?

**Reply by reyemtm** [reply]
> You're testing two completely different things. An mbtiles file is:
    Pre-projected to Web Mercator
    Pre-sliced into XYZ tiles
    A static file

I guess that is where I get a bit confused, as the COG is a static file, has internal pyramids which are similar I assume to tiles, and pre-projected to Web Mercator. For more answers to the other questions and comments I will post to the main thread.

**Reply by vincentsarago** [reply]
see: https://gdal.org/drivers/raster/cog.html#reprojection-related-creation-options 

with rio-cogeo you can use do: `rio cogeo create in.tif cog.tif -w --use-cog-driver`

**Comment by vincentsarago** [comment]
1-2 seconds is quite slow, I would expect something around 200ms 
question though:
- is your COG stored locally (next to your server?)
- is the COG WebOptimized? (meaning aligned with WebMercator grid)
- do you limit the number of concurrency per thread? 

TiTiler is not meant to be compared with mbtile tile server or other server working with locally stored tiles. It will never be faster than those because TiTiler (gdal/rasterio) has to do more work. if you still want to work with COG files but want something fast you should look at https://github.com/blacha/cogeotiff.

TiTiler offers more than just basic tile serving. The main goal is to enable `dynamic` processing (e.g rescaling, colormap) so we have to accept some latency. 

The major problem of TiTiler is GDAL. Every Tile read will block the thread because GDAL is a blocking process, sadly we are not able to `fix` this.

There is no `benchmarks` available because the performance depends also on the type of file you will use as input. That is said that could be an interesting thing to do

**Reply by vincentsarago** [reply]
if you want to understand a bit more what's going on you can checkout https://github.com/developmentseed/tilebench but also https://developmentseed.org/titiler/advanced/performance_tuning/

**Reply by reyemtm** [reply]
I did fing that cogeotiff repo but in my limited testing I was not able to get a tile server running based on their examples. Maybe I should check it out again. From what I recall there is no pre-built COG server, but the pieces are there...

**Comment by reyemtm** [comment]
I get that TiTiler and GeoTIFF servers in general are not specifically designed to be an XYZ tile server to replace traditional solutions, my question is whether it *can* and then if it *should* be used for this purpose. Why - mainly due to how the ortho imagery is stored - COG on very cheap object storage - vs mbtiles on slightly more costly block storage. 

And - the presenter from Bayer who just talked at FOSS4G said he is doing just this - they have replaced a cached tiling solution with rio-tiler and find it *more* performant, but he did not get into details in regards to the CPU infrastructure except that it is running on a cluster with auto-scaling.

**Reply by vincentsarago** [reply]
To be honest it's hard for us to help you because we don't know:
- the configuration of your server (see https://developmentseed.org/titiler/advanced/performance_tuning/)
- the information about the files
🤷

**Comment by reyemtm** [comment, marked as answer]
After some digging I'll answer my own question. As can be seen from the comments above the answer is nuanced, but the short answer is Yes, a COG server can be used in place of an mbtiles server for serving out simple ortho image tiles. Latency or response times can be expected to be under 500ms given adequate hardware, properly formed TIFFs. https://github.com/linz/basemaps is probably more suited to this specific use case than either TiTiler of Terracotta...but all these projects are still a work in progress.
